Lib/fractions.py

```diff
@@ -13,6 +13,12 @@
 __all__ = ['Fraction']
 
 
+def _gcd(a, b):
+    # Supports non-integers for backward compatibility.
```

Comment:
My main worry with this PR is this comment. I don't think that we should keep it back "backward compatibility". Either we remove _gcd() or we keep it on deliberate purpose, to return the same type than inputs (a and b arguments).

I understand that @mdickinson wants to return the same type, and I'm fine with that. But in this case, the comment should be something like:

```
For int sublcasses and other types, math.gcd(a, b) call the __index__() method on arguments
and it always return the int type. This function is used to return the same type than input.
```

Reply:
Maybe also add a comment in Fraction constructor to explain why we test type() == int: explain that we want to get g as the same type than numerator and denominator types. Or something like that ;-)

Reply:
@vstinner This comment wasn't introduce in this PR; it's from the previously working code.

> I understand that @mdickinson wants to return the same type

No.  I simply want to undo the breakage that #18021 introduced, in the simplest way possible, and with a regression test. I'm open to discussing any sort of behaviour change after that, but the first priority should be to get back to a working state, and the simplest path is to get back to the working state that existed in 3.9.0a2.

Another way to address the breakage would be to revert #18021 in its entirety, and then re-do it more carefully.

Reply:
The statistics module tries to keep input types for the output type. I think that it's good thing for generic modules like fractions and statistics. The documentation seems to say that it keeps the input type, but it's not *very* explicit: https://docs.python.org/dev/library/fractions.html

I'm fine with keeping Python 3.8 behavior (so use your PR). If someone wants to change it, it can be done later.

My first intent was only to remove the public deprecated fractions.gcd() function. I don't care about private functions like fractions._gcd().